R/utils-get_code_dependency.R

```diff
@@ -0,0 +1,267 @@
+#' objects that are modified or created by function calls with side effects, a `@linksto` comment tag was introduced.
+#' This tag enables to restore the reproducible code for an object, with all side effects that do not directly change
+#' this object. Read more about side effects and the usage of `@linksto` tag in [`get_code_dependencies()`] function.
+#'
+#' @param calls_pd A `list` of `data.frame`s, which is a result of `utils::getParseData()` grouped into separate calls.
+#' A result of `extract_calls()` function.
+#' @return A `list` of length equal to the lenght of `calls_pd` of character vectors of names of objects that are
+#' influenced by `@linksto` tag in a corresponding `call` element of `calls_pd`.
+#' @keywords internal
+#' @noRd
+extract_side_effects <- function(calls_pd) {
+  lapply(
+    calls_pd,
+    function(x) {
+      linksto <- grep("@linksto", x[x$token == "COMMENT", "text"], value = TRUE)
+      unlist(strsplit(sub("\\s*#\\s*@linksto\\s+", "", linksto), "\\s+"))
+    }
+  )
+}
+
+#' @keywords internal
+#' @noRd
+remove_graph_duplicates <- function(graph) {
+  # This function is just a nice to have.
+  # This is not crucial for the implementation.
+  # Purpose: remove duplicates before ":" and after, in each element of list.
+
+  lapply(
+    lapply(graph, paste, collapse = "***"),
+    function(x) {
+      split <- strsplit(x, split = ":", fixed = TRUE)[[1]]
+      deps <- strsplit(split[1], split = "***", fixed = TRUE)[[1]]
+
+      if (length(split) == 1) {
+        unique(deps)
+      } else if (length(split) == 2) {
+        infl <- strsplit(split[2], split = "***", fixed = TRUE)[[1]][-1]
+        c(unique(deps), ":", unique(infl))
+      }
+    }
+  )
+}
+
+# graph_parser ----
+
+#' Return the indexes of calls of code needed to reproduce the object
+#'
+#' @param x The name of the object to return code for.
+#' @param graph A result of `code_graph()`.
+#' @param skip `NULL` or `character` vector. In a recursive call, it is needed to drop parent object to omit dependency
+#' cycles.
+#'
+#' @return `numeric` vector indicating which calls of `graph` are required to build the object passed by name in `x`.
+#'
+#' @keywords internal
+#' @noRd
+graph_parser <- function(x, graph, skip = NULL) {
+  skip <- c(x, skip)
+
+  occurence <-
+    vapply(
+      graph, function(call) {
+        if (":" %in% call) {
+          call <- call[1:(which(":" == call) - 1)]
+        }
+        x %in% call
+      },
+      logical(1)
+    )
+
+  influencers <-
+    unlist(
+      lapply(graph[occurence], function(call) {
+        if (":" %in% call) {
+          call[(which(":" == call) + 1):length(call)]
+        }
+      })
+    )
+  influencers <- setdiff(influencers, skip)
+
+  if (length(influencers)) {
+    influencers_ids <-
+      lapply(influencers, function(influencer) {
+        graph_parser(influencer, graph[1:max(which(occurence))], skip)
+      })
```

Comment:
Is `skip` indispensible?
```suggestion
      lapply(setdiff(influencers, x), function(influencer) {
        graph_parser(influencer, graph[1:max(which(occurence))])
      })
```

Reply:
yes, because it keeps getting extended in every loop.

```r
a <- 1
b <- a
c <- b
a <- c
```

and if you don't have skip, then you would end up in the infinite loop

Reply:
And the `setdiff` is not enough?

Reply:
Hmm, let's consider below flow

We seek for code of `a`. 
1st iteration: 'x <- "a"` and `influencers <- "c"` and then we run `graph_parser` recursively for `x <- "c"` and `graph` is reduced to `1:4` (as `a` exists in line 4)
2nd iteration: `x <- "c"` and `influencers <- "b"` and then we run `graph_parser` recursively for `x <- "b"` and `graph` is 
 reduced to `1:4` (as `c` exists in line 4)
3rd iteration: `x <- "b"` and `influencers <- "a"` and again you run `graph_parser` recursively for `x <- "a"` and `graph` is reduced to `1:3` (as `b` exists in line 3)

Ok, so I think we no longer arrive at `a` being influenced by `c` again since the graph got reduced. Will try with `setdiff`

Reply:
Thanks for the valuable feedback - just removed `skip` parameter 0891a1f8c52ced3a81984add5a95a849d8876ba5 and added a test that detects if we fall in a loop